### PR TITLE
IDataType generic restriction

### PR DESCRIPTION
**Pull request by pan3793**
Partial implement #247

### What changes in this PR?
- Move classes out of the package `jdbc` except JDBC relevant
- Refactor `IDataType` with Generic Type constraint
- Split `DataTypeIntX` into `DataTypeUIntX` and `DataTypeIntX`
- Now ClickHouse type mapping `IDataType` 1:1
- Move type cast logic out from `IDataType`
- Distinguish `javaType` and `jdbcJavaType` in `IDataType`
- Misc code refactor

### CODE REVIEW

**Comment by pan3793** [comment]
retest this please

**Comment by pan3793** [comment]
> retest this please

Oops, it doesn't work.

File: clickhouse-native-jdbc/src/main/java/com/github/housepower/data/DataTypeFactory.java [hunk 1]
```diff
@@ -114,29 +106,30 @@ public static IDataType get(SQLLexer lexer, NativeContext.ServerContext serverCo
         registerType(creators, new DataTypeInt32("Int32"));
         registerType(creators, new DataTypeInt64("Int64"));
 
-        registerType(creators, new DataTypeInt8("UInt8"));
-        registerType(creators, new DataTypeInt16("UInt16"));
-        registerType(creators, new DataTypeInt32("UInt32"));
-        registerType(creators, new DataTypeInt64("UInt64"));
+        registerType(creators, new DataTypeUInt8("UInt8"));
```
Comment (sundy-li) [inline clickhouse-native-jdbc/src/main/java/com/github/housepower/data/DataTypeFactory.java:109]:
```suggestion
        new DataTypeInt32()
```

Comment (pan3793) [reply clickhouse-native-jdbc/src/main/java/com/github/housepower/data/DataTypeFactory.java]: Updated

**Comment by sundy-li** [comment]
LGTM， nice pr.

**Pull request merged**